src/commands/sourcemaps/git.ts

```diff
@@ -0,0 +1,242 @@
+
+// Returns the tracked files of the current repository.
+export const gitTrackedFiles = async (git: simpleGit.SimpleGit): Promise<string[]> => {
+  const files = await git.raw('ls-files')
+
+  return files.split(/\r\n|\r|\n/)
+}
+
+// Trims from a set of characters from the start of a string.
+const trimStart = (str: string, chars: string[]) => {
+  let start = 0
+  const end = str.length
+  while (start < end && chars.indexOf(str[start]) >= 0) {
+    ++start
+  }
+
+  return start > 0 ? str.substring(start, end) : str
+}
+
+// Trims from a set of characters from a string.
+const trim = (str: string, chars: string[]) => {
+  let start = 0
+  let end = str.length
+  while (start < end && chars.indexOf(str[start]) >= 0) {
+    ++start
+  }
+  while (end > start && chars.indexOf(str[end - 1]) >= 0) {
+    --end
+  }
+
+  return start > 0 || end < str.length ? str.substring(start, end) : str
+}
+
+// Generates a proper source file path from a sourcemap:
+// - Strip a set of hard-coded prefixes ('webpack:///./')
+// - Strip the eventual projectPath
+// - Removes query parameters
+//
+// It returns the new source as well as whether or not the specified projectPath was stripped.
+//
+// For example, the following source path:
+// webpack:///./project/folder1/folder2/src.js?abc123
+//
+// Will be cleaned up into 'project/folder1/folder2/src.js'.
+export const cleanupSource = (source: string, projectPath: string): [string, boolean] => {
+  // Prefixes
+  const prefixesToRemove = ['webpack:']
+  for (const p of prefixesToRemove) {
+    if (source.startsWith(p)) {
+      source = source.slice(p.length)
+    }
+  }
+  source = trimStart(source, ['/', '.'])
+  // ProjectPath
+  projectPath = trim(projectPath, ['/', '.'])
+  const projectPathFound = source.substr(0, projectPath.length) === projectPath
+  if (projectPathFound) {
+    source = source.slice(projectPath.length)
+  }
+  // Query parmeter
+  const pos = source.lastIndexOf('?')
+  if (pos > 0) {
+    source = source.slice(0, pos)
+  }
+
+  return [trimStart(source, ['/', '.']), projectPathFound]
+}
+
+// Creates a lookup map from source paths to tracked file paths.
+// The tracked file paths will be split into as many keys as there is folders in the path (+ the filename).
+// It allows to do an exact match when looking up a source path in the tracked files, which will be enough
+// in most cases.
+// It also allows to make an 'educated guess' / 'best-effort' match on the off-chance the source path we are
+// trying to find does not contain the whole tracked file path. It's an edge case but can easily be handled
+// using this method.
+//
+// To perform a match simply lookup the source path in the map keys. The value returned is the complete tracked
+// file path.
+//
+// For example the following tracked file path:
+// project/folder1/folder2/src.js
+//
+// Will be matched by any of:
+// project/folder1/folder2/src.js
+// folder1/folder2/src.js
+// folder2/src.js
+// src.js
```

Comment:
Yes it could allow some false positives because of the effort to create a match whenever possible. In the example you described `src.js` would indeed be resolved to the `lib/src.js` tracked file.

What you are saying is that we should only have an entry `src.js` => `lib/src.js` if the sourcemap we're currently uploading has `lib` as a prefix for all of the 'sources'?

To boil the issue down to two cases:

- `webpack:///./project/folder/src.js` with `folder/src.js` as tracked file => 
    To have a proper match we ask the user to specify in the `--project-path` flag the `project` value.

- `webpack:///./folder/src.js` with `project/folder/src.js` as tracked file (this is the monorepo case for example) =>
    To have a proper match we don't ask anything but try the `trackedFilesMap` method of removing folders from the tracked files paths.

Assuming the second case only ever happens in a monorepo-like case, then indeed all sources of a sourcemap should have the same prefix to remove. It's tricky tbh.

We could also require the user to provide yet another flag to cover the second case and therefore have zero missing information and just do perfect matches. wdyt?